### PR TITLE
PTY session crash detection and reporting

### LINKED ISSUE #147: PTY sessions need crash recovery and status reporting
## Description

When a PTY/tmux session crashes (e.g., Claude exits, node-pty spawn failure, tmux session killed), the app has no indication that the session is dead. The user sees a frozen terminal with no error message or recovery option.

## Observed Behavior

User attached to a tmux session (exodus-loop) via auto-discovery. The session crashed silently — switching to that tab in the app showed nothing useful.

## Proposed Fix

- Detect PTY session exit/crash in PtyManager and emit a session error event
- Surface session health status in the SessionPicker (e.g., red dot for crashed sessions)
- Show an error banner in the session view when the backing process is dead
- Consider auto-removing or offering to restart crashed sessions

## Acceptance Criteria

- [ ] Crashed PTY sessions show an error state in the UI
- [ ] User can dismiss or restart a crashed session
- [ ] SessionPicker indicates session health status

### PR DESCRIPTION
**Pull request by blamechris**
## Summary

Implements crash detection and status reporting for PTY sessions to address issue #147.

- Adds periodic health checks (every 30s) to detect when Claude process crashes inside tmux
- Emits `session_crashed` event through the session management chain
- Reports crashes to connected clients via `session_error` WebSocket message
- Properly cleans up health check timers on session destroy

### PtyManager (`pty-manager.js`)
- Added `_healthCheckInterval` and `_healthCheckIntervalMs` (30s) properties
- Added `_startHealthCheck()` to start periodic health checks on session start
- Added `_stopHealthCheck()` to clean up timers on destroy
- Added `_checkHealth()` to verify tmux session exists and panes are alive
- Emits `crashed` event with reason when health check fails

### PtySession (`pty-session.js`)
- Listens for `crashed` events from PtyManager
- Emits `session_crashed` event to SessionManager with error details

### SessionManager (`session-manager.js`)
- Wires up `session_crashed` event from PtySession instances
- Emits crash events with sessionId for WebSocket broadcast

### WsServer (`ws-server.js`)
- Listens for `session_crashed` events from SessionManager
- Broadcasts `session_error` message to clients on affected session

## Health Check Strategy

The implementation uses tmux's built-in pane status monitoring:
- Every 30s, checks if tmux session exists via `tmux has-session`
- Queries pane status via `tmux list-panes -F '#{pane_dead}'`
- Emits crash event if session missing or panes are dead
- Stops health checks after detection to prevent repeated alerts

## Test Plan

- [x] Existing tests pass
- [ ] Manual test: Start PTY session, kill Claude process, verify crash detection after 30s
- [ ] Manual test: Verify WebSocket clients receive `session_error` message
- [ ] Manual test: Verify timer cleanup on session destroy (no memory leaks)

Closes #147

### CODE REVIEW

**Comment by blamechris** [comment]
## Agent Review Summary

Reviewed PR #243 against project conventions, architecture, code quality, security, and completeness criteria. Overall implementation is solid with good architecture decisions, but there are some issues to address.

---

## Critical Issues

### 1. Unrelated Change in MarkdownRenderer
The PR includes a change to `/packages/app/src/components/MarkdownRenderer.tsx` that reduces list indentation (4 spaces → 3 spaces, capped at 3 levels). This is completely unrelated to crash detection and should be in a separate PR.

**Action:** Remove this change from the PR. If it's a UX improvement worth keeping, create a dedicated PR for it.

---

## Architectural Concerns

### 2. Auto-Discovery Feature Scope Creep
The PR adds a significant new feature (auto-discovery of new tmux sessions) that wasn't mentioned in the PR description or issue #147. This includes:
- `autoDiscovery` flag and `discoveryIntervalMs` config
- `startAutoDiscovery()` and `stopAutoDiscovery()` methods
- `_pollForNewSessions()` polling logic
- `new_sessions_discovered` event emission
- `_lastDiscoveredSessions` tracking

**Issues:**
- Auto-discovery is orthogonal to crash detection
- No documentation on when/why this should be enabled
- No indication of whether this is enabled by default in either server mode
- No WebSocket forwarding of `new_sessions_discovered` events (they're emitted but never consumed)
- No client-side handling of these events

**Action:** Either:
1. Remove auto-discovery from this PR and create a separate PR with proper design/documentation
2. If it's essential for crash recovery UX, document the rationale and wire it up end-to-end

---

## Code Quality Issues

### 3. Missing Test Coverage
No tests exist for PtyManager. This PR adds critical crash detection logic but no tests to verify:
- Health check timers start/stop correctly
- Crash events are emitted with correct payloads
- Timer cleanup prevents memory leaks
- Edge cases (tmux command failures, empty pane lists, etc.)

**Action:** Add unit tests for PtyManager, especially the new health check methods.

### 4. Magic Number Without Justification
30-second health check interval is hardcoded. Why 30s? Why not configurable?

**Recommendation:** Add a comment explaining the choice or make it configurable via constructor.

### 5. Incomplete PR Test Plan
Test plan says:
- [x] Existing tests pass
- [ ] Manual test: Start PTY session, kill Claude process, verify crash detection after 30s
- [ ] Manual test: Verify WebSocket clients receive session_error message
- [ ] Manual test: Verify timer cleanup on session destroy

**Action:** Complete manual testing and check off items before merge.

---

## Security/Robustness Concerns

### 6. Shell Injection Risk (Low)
Line 161 in pty-manager.js:
```javascript
const output = execSync(`tmux list-panes -t ${this.sessionName} -F '#{pane_dead}' 2>/dev/null`, {
```

While `sessionName` is validated elsewhere (attach_session checks for valid tmux names), it's good practice to avoid string interpolation in shell commands.

**Recommendation:** Use an allow-list pattern or explicit validation at construction time.

### 7. Missing Error Boundary
If `_checkHealth()` throws unexpectedly, the timer keeps running and calling a broken method every 30s.

**Current mitigation:** The try/catch in `_checkHealth()` handles this and calls `_stopHealthCheck()`.

**Status:** Acceptable, but worth noting for future robustness.

---

## Minor Issues

### 8. Inconsistent Error Handling
In `_checkHealth()`, the catch block emits `crashed` with `reason: 'health_check_error'` and `error: err.message`. But the non-error crash reasons (`session_not_found`, `pane_dead`) don't include an `error` field consistently.

In pty-session.js line 111:
```javascript
const errorMsg = error ? `${reason}: ${error}` : reason
```

This works, but it's a bit defensive. Consider normalizing the payload structure.

**Recommendation:** Always include `error` field (even if null) for consistency.

### 9. Event Chain Verbosity
The event flow is clear but verbose:
1. PtyManager emits `crashed`
2. PtySession listens, emits `session_crashed`
3. SessionManager listens, emits `session_crashed` again
4. WsServer listens, broadcasts `session_error`

**Status:** This is fine for maintainability and follows the existing architecture pattern. Not a blocker.

---

## Positive Notes

**What went well:**
- Clean separation of concerns (health check in PtyManager, event wiring in PtySession/SessionManager)
- Follows existing EventEmitter patterns consistently
- Stops health checks on crash to avoid repeated alerts
- Properly cleans up timer in destroy()
- WebSocket message structure matches existing `session_error` format
- Logging is helpful for debugging

---

## Acceptance Criteria vs Issue #147

Original issue requested:
- [ ] Crashed PTY sessions show an error state in the UI → **Partial**: `session_error` is sent but no follow-up on UI changes
- [ ] User can dismiss or restart a crashed session → **Not implemented**
- [ ] SessionPicker indicates session health status → **Not implemented**

**This PR only implements backend crash detection.** The UI improvements are missing.

**Action:** Either:
1. Update PR description to clarify this is "Phase 1: Backend detection only"
2. Add follow-up issue for UI work
3. Keep issue #147 open and update it to reflect that only detection is done

---

## Recommendation

**DO NOT MERGE** until:
1. MarkdownRenderer change is removed
2. Auto-discovery feature is either removed or properly documented/wired up
3. Manual test plan is completed
4. A follow-up issue is created for UI work (or #147 is updated to reflect partial completion)

**Consider adding:**
- Unit tests for PtyManager health checks
- Config option for health check interval
- Client-side crash recovery UI (separate PR)

---

**Review completed:** 2026-02-09
**Agent:** Claude Opus 4.6 via Sonnet 4.5

**Comment by blamechris** [comment]
## Follow-Up Issues Created

Based on the review above, I've created the following issues for out-of-scope work:

- **#246**: Add unit tests for PtyManager health checks
- **#247**: UI crash recovery for PTY sessions (SessionPicker indicators, restart/dismiss actions)
- **#248**: Clarify/remove auto-discovery feature in SessionManager

---

## Summary of Required Changes Before Merge

1. **Remove MarkdownRenderer change** (unrelated to crash detection)
2. **Address auto-discovery** (remove or document/wire up per #248)
3. **Complete manual test plan** (check off the 3 manual test items in PR description)
4. **Update issue #147** to clarify this is "Phase 1: Backend detection" (UI work tracked in #247)

Once these are addressed, the core crash detection implementation is solid and ready to merge.

File: packages/server/src/session-manager.js [hunk 1]
```diff
@@ -99,6 +100,10 @@ export class SessionManager extends EventEmitter {
     this._messageHistory = new Map() // sessionId -> Array<{ type, ...data }>
     this._pendingStreams = new Map() // sessionId:messageId -> accumulated delta text
     this._maxHistory = 100
+    this._autoDiscovery = autoDiscovery
+    this._discoveryIntervalMs = discoveryIntervalMs
+    this._discoveryTimer = null
+    this._lastDiscoveredSessions = new Set() // Track tmux session names we've seen
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/src/session-manager.js]: **FALSE POSITIVE**

**Reason:** This is a known artifact from PR cross-contamination during rebase. The auto-discovery code (`autoDiscovery`, `discoveryIntervalMs`, `_discoveryTimer`, `_lastDiscoveredSessions`) is tracked separately in issue #138 and will be moved to its own PR.

**Evidence:** 
- PR title/description: "PTY session crash detection and reporting" — no mention of auto-discovery
- Issue #138 explicitly tracks auto-discovery as a separate feature
- The crash detection functionality (health checks, `session_crashed` event) is independent and works without these fields

The `session_crashed` wiring in `_wireSessionEvents()` is the only crash-detection-specific addition to SessionManager. The auto-discovery fields will be removed in a follow-up rebase or addressed in PR for #138.
